### PR TITLE
Re-vamp Bloom filter and sparse-transforms

### PR DESCRIPTION
**Pull request by nevillelyh**
Depends on #2635 

- Use new BF API for all sparse transforms.
- By default split large BFs into 100MB ones and get rid of warnings.
- Set BF capacity to match input size when running in tests.

### CODE REVIEW

**Comment by anish749** [comment]
I am a bit skeptical about Funnel being exposed as an implicit directly. I don't have a good fix.
 - We won't be able to use vendored Guava. 
 - Users would now need to explicitly add import from magnolify.

 - We are exposing `ABloomFilter`. How would users use this?
 - Should we have `PartitionSettings` as a dependent type inside the `BloomFilter`? Considering that if we plan to expand this API to cover filters other than BloomFilters we might need to change the PartitionSettings for those filters.

**Comment by nevillelyh** [comment]
- We can use a shim, but `ApproxFilterCompanion#Hash` already does that. So you can technically have a `VendoredBloomFilter`. OTOH none of this can survive Guava BF breaking binary encoding (Beam may bump vendered Guava) so feels like a moot point. Like-wise even though we avoid `Funnel` going through Java ser/de which might break binary due to Magnoia changes, it does not protect us against `Funnel` behavior change which breaks correctness. Basically there's no way against binary encoding or funnel changes so we're relying on them not breaking.
- They don't if we include basic implicits in the namespace, e.g. `Funnel[String]`, `Funnel[Array[Byte]]`. Magnolify is only needed for composites.
- It's just an illustration how the interface could work with non-Guava/Funnel based impl, as mentioned in https://github.com/spotify/scio/pull/2635#issuecomment-580469808. I plan to remove it before merging.
- Or we can make `PartitionSettings` private so it's used in `hash*` transforms only and limited to BF.

As I said repeatedly the goal is to simplify things not scope creeping every time a minor issue surfaces.

**Comment by anish749** [comment]
Sounds great.
I am not thinking about serde. So it is fine. I don't expect much Guava dependency issues 🤞.

👍 to including funnel of primitive types in the namespace / package level.
👍 to making `PartitionSettings` private.

**Comment by nevillelyh** [comment]
Actually I'm not sure about including `Funnel` implicits. The problem is we can't add a `object Funnel` so no matter where we put the implicts, they WILL have to be imported explicitly, so we might as well instruct users to `import magnolify.guava.auto._`.

**Comment by nevillelyh** [comment]
Did some benchmark. 1 billion input UUIDs to construct BFs, 500 million UUIDs as negatives. 8 n1-standard-4 workers, auto-scaling off. This results in 12 BFs of ~100MB each at `fpp = 0.01`

**Guava BF**
Elapsed time: 30 min 16 sec
Total vCPU time: v14.651 vCPU hr
Total memory time: 54.943 GB hr
Total PD time: 915.723 GB hr

Will update the baseline one once it finishes (currently at 41min & still running).

**Comment by nevillelyh** [comment]
This is ready for review. It does break source compatibility though and requires explicit `import magnolify.guava.auto._`.

To avoid the extra import, we'll have to change magnolify & introduce a wrapper for `Funnel[T]`, and move implicits in its companion object. Thoughts?

**Comment by anish749** [comment]
I was thinking about the magnolify funnel imports, and I came up with this..

Goal:
 - we don't want to add the magnolify import for primitive types, as that would break the api.
 - We still want to use magnolify for types that need derivation.

I haven't tested this, but may be this should work.
We create a new type class `SFunnel` and have to / from methods to Guava funnel.

```scala
trait SFunnel[T] {

  def funnel(from: T, sink: PrimitiveSink): Unit

}

object SFunnel {

  implicit val integerSFunnel: SFunnel[Int] = Funnels.integerFunnel().asInstanceOf[Funnel[Int]]

  implicit def fromGuava[T](gFunnel: Funnel[T]): SFunnel[T] = {
    new SFunnel[T] {
      override def funnel(from: T, sink: PrimitiveSink): Unit = gFunnel.funnel(from, sink)
    }
  }

  def toGuava[T](sfun: SFunnel[T]): Funnel[T] = {
    new Funnel[T] {
      override def funnel(from: T, into: PrimitiveSink): Unit = sfun.funnel(from, into)
    }
  }
}
```

Usage:
```
val funnel: Funnel[T] = SFunnel.toGuava(implicitly[SFunnel[T]])
```
Downside, we call two extra functions in the call stack.

**Comment by nevillelyh** [comment]
I thought about that, but we'll still need a bunch of `implicit val`s in the companion, which duplicates ones in `magnolify.guava.auto._`. Need to double check if they cause any ambiguity.

**Comment by jto** [comment]
Not a fan of those implicit conversions. In my experience that's the kind of stuff that causes super weird issues 6 month later...

File: scio-core/src/main/scala/com/spotify/scio/hash/ApproxFilter.scala [hunk 1]
```diff
@@ -0,0 +1,300 @@
...
+   * will result in its saturation, and a sharp deterioration of its false positive probability.
+   */
+  final def create[T: Hash](
+    elems: SCollection[T],
+    expectedInsertions: Long,
+    fpp: Double
+  ): SCollection[Filter[T]] = {
+    implicit val elemCoder = Coder.beam(elems.internal.getCoder)
+    elems.transform {
+      _.groupBy(_ => ()).values
```
[the remark this answers is not in the document]
Comment (nevillelyh) [reply scio-core/src/main/scala/com/spotify/scio/hash/ApproxFilter.scala]: Actually no, with the most typical use case of bounding BF raw bitset to 100MB, at `fpp` 0.01 we can fit roughly 80 million `elems` per group, assuming 36 byte UUID that's < 3GB in a lazy `Iterable`. That's pretty small with modern shuffle capability. In practice it's actually faster than creating many 100MB bit sets and merging them (bitwise OR), especially considering the ser/de required for BFs across workers.

**Comment by nevillelyh** [comment]
I fixed everything except the import requirement, which I can live with considering the extra complexity of adding more implicits.

Should we keep this as is for now and merge once we're ready for 0.9.x branch?

**Pull request merged**